src/main/java/org/xbill/DNS/lookup/DnsSession.java

```diff
@@ -0,0 +1,214 @@
+public class DnsSession {
+  private static final int DEFAULT_MAX_ITERATIONS = 16;
+  public static final int DEFAULT_NDOTS = 1;
+  /** The {@link Resolver} to use to look up records. */
+  @NonNull private final Resolver resolver;
+  /**
+   * Sets the maximum number of CNAME or DNAME redirects allowed before lookups with fail with
+   * {@link RedirectOverflowException}.
+   */
+  @Builder.Default private final int maxRedirects = DEFAULT_MAX_ITERATIONS;
+  /**
+   * Sets the threshold for the number of dots which much appear in a name before it is considered
+   * absolute. The default is {@code 1}, meaning meaning that if there are any dots in a name, the
+   * name will be tried first as an absolute name.
+   */
+  @Builder.Default private final int ndots = DEFAULT_NDOTS;
+
+  @Singular("searchPath")
+  private final List<Name> searchPath;
+
+  // this augments the lombok generated code.
+  public static class DnsSessionBuilder {
+    public DnsSession build() {
+      // note that this transform is idempotent, as concatenating an already absolute Name with root
+      // is a noop.
+      if (searchPath != null) {
+        this.searchPath =
+            searchPath.stream()
+                .map(
+                    name -> {
+                      try {
+                        return Name.concatenate(name, Name.root);
+                      } catch (NameTooLongException e) {
+                        throw new IllegalArgumentException("Search path name too long");
+                      }
+                    })
+                .collect(Collectors.toCollection(ArrayList::new));
+      }
+      return innerBuild();
+    }
+  }
+
+  /**
+   * Make an asynchronous lookup of the provided name.
+   *
+   * @param name the name to look up.
+   * @param type the type to look up, values should correspond to constants in {@link Type}.
+   * @param dclass the class to look up, values should correspond to constants in {@link DClass}.
+   * @return A {@link CompletionStage} what will yield the eventual lookup result.
+   */
+  public CompletionStage<LookupResult> lookupAsync(Name name, int type, int dclass) {
+    CompletableFuture<LookupResult> future = new CompletableFuture<>();
+    lookupUntilSuccess(expandName(name).iterator(), type, dclass, future);
+    return future;
+  }
+
+  /**
+   * Generate a stream of names according to the search path application semantics. The semantics of
+   * this is a bit odd, but they are inherited from Lookup.java. Note that the stream returned is
+   * never empty, as it will at the very least always contain name.
+   */
+  Stream<Name> expandName(Name name) {
+    if (name.isAbsolute()) {
+      return Stream.of(name);
+    }
+    if (name.length() == 255) {
+      // handle this special case where the name is relative but can not be made absolute as it
+      // is at it's maximum length and nothing can be appended.
+      throw new IllegalArgumentException("Name is too long");
```

Comment:
Can't you use `NameTooLongException`?

Reply:
Since `NameTooLongException` is a checked exception it would add an exception to the `lookupAsync` and one for a pretty esoteric case as well. (Since it takes a Name and names can't be constructed if they are too long, it is only for the case where it is a maximally long relative name that can not be made absolute).

I think this is a good use case for a checked exception, as it specifies exactly what is going on for the edge case but for the common case it can be ignored and the failure will bubble up the call stack. Callers that want to deal with it can do that, but callers that don't doesn't need to.

Moreover, since I envision lookupAsync() to be used in chains of CompletionStage based async code, having it throw a checked exception would make those chains a lot more verbose as it would need to be either handled out of the call stack or wrapped in an unchecked exception.

Reply:
Come to think of it I think that the better way to deal with this is to change Name to fail creation for relative names of length 255. That way the caller would need to handle that the same way as it would need to handle other Name creations that are much longer.

Reply:
Without following the code too much now, what happens with two search paths where only the second will generate a NTLE? If that is communicated as a failed future, you could do the same with with the first path (by try/catching expandName at the call site and immediately returning a failed future. No need for a checked exception in the signature of lookupAsync.

As for limiting the Name length, not sure. Let's discuss this somwhere else.

Reply:
Ah, I did not think of that. Having any checked exception be passed on through the CompletionStage is definitely one way of doing it.

I'm okay with that, although I think that changing the Name constructor to disallow it is the better tradeoff, but we can discuss that in the PR that I just opened. 

My fairly strong opinion is that checked exceptions as designed in Java is a flawed concept, but that is not something that we will fix tonight.

Reply:
The concept is broken, yes. Especially now with the lambdas. But Java has been born disabled. Signed bytes. Sigh.